src/base/matrix.rs

```diff
@@ -1855,68 +1880,94 @@ macro_rules! impl_fmt {
-                        max_length = crate::max(max_length, val_width(&self[(i, j)], f));
+                    for i in 0..nrows {
+                        for j in 0..ncols {
+                            max_length = crate::max(max_length, val_width(&self[(i, j)], f));
+                        }
                     }
-                }
 
-                let max_length_with_space = max_length + 1;
-
-                writeln!(f)?;
-                writeln!(
-                    f,
-                    "  ┌ {:>width$} ┐",
-                    "",
-                    width = max_length_with_space * ncols - 1
-                )?;
-
-                for i in 0..nrows {
-                    write!(f, "  │")?;
-                    for j in 0..ncols {
-                        let number_length = val_width(&self[(i, j)], f) + 1;
-                        let pad = max_length_with_space - number_length;
-                        write!(f, " {:>thepad$}", "", thepad = pad)?;
-                        match f.precision() {
-                            Some(precision) => {
-                                write!(f, $fmt_str_with_precision, (*self)[(i, j)], precision)?
+                    let max_length_with_space = max_length + 1;
+
+                    writeln!(f)?; // leading newline to ensure no offset from previous prints
+                    writeln!(
+                        f,
+                        "  ┌ {:>width$} ┐",
+                        "",
+                        width = max_length_with_space * ncols - 1
+                    )?;
+
+                    for i in 0..nrows {
+                        write!(f, "  │")?;
+                        for j in 0..ncols {
+                            let number_length = val_width(&self[(i, j)], f) + 1;
+                            let pad = max_length_with_space - number_length;
+                            write!(f, " {:>thepad$}", "", thepad = pad)?;
+                            match f.precision() {
+                                Some(precision) => {
+                                    write!(f, $fmt_str_with_precision, (*self)[(i, j)], precision)?
+                                }
+                                None => write!(f, $fmt_str_without_precision, (*self)[(i, j)])?,
                             }
-                            None => write!(f, $fmt_str_without_precision, (*self)[(i, j)])?,
                         }
+                        writeln!(f, " │")?;
                     }
-                    writeln!(f, " │")?;
-                }
 
-                writeln!(
-                    f,
-                    "  └ {:>width$} ┘",
-                    "",
-                    width = max_length_with_space * ncols - 1
-                )?;
-                writeln!(f)
+                    write!(
+                        f,
+                        "  └ {:>width$} ┘",
+                        "",
+                        width = max_length_with_space * ncols - 1
+                    )
```

Comment:
Previously this added two newlines after the end of the matrix. i.e. [ 1.0 ]/n/n
This removes both of those giving flexibility to write immediately after the matrix. Could add one or both of the newlines back in if we don't want to change it too much.

```rust
println!("Display");
println!("{}directly after", smat);
println!("next line");
```

Old:
```
Display

  ┌         ┐
  │ 1 2 3 4 │
  │ 5 6 7 8 │
  └         ┘

directly after
next line
```

New:
```
Display

  ┌         ┐
  │ 1 2 3 4 │
  │ 5 6 7 8 │
  └         ┘directly after
next line
```

Reply:
This looks like the right decision to me.

Is the initial newline necessary though? I guess it is in order to give reasonable formatting for cases where you have e.g.
```rust
println!("My matrix is {matrix}");
```
I guess if you don't have a leading newline, cases like this will end up with messed up formatting?

Reply:
Yeah I assume that is why the initial newline is included. I wouldn't be against removing it myself.